File: Kattis/chess.cpp
#include <iostream>
#define MIN( a, b ) ( ( (a) < (b) ) ? (a) : (b) )
#define MAX( a, b ) ( ( (a) > (b) ) ? (a) : (b) )

using namespace std;

int main() {

	int cases, x1, y1, x2, y2;
	char c1, c2;

	cin >> cases;

	while ( cases-- ) {

		cin >> c1 >> y1 >> c2 >> y2;

		x1 = c1 - 'A';
		x2 = c2 - 'A';

		if ( ( x1 + y1 ) % 2 != ( x2 + y2 ) % 2 )
			cout << "Impossible" << endl;
		else if ( x1 == x2 && y1 == y2 )
			cout << 0 << ' ' << c1 << ' ' << y1 << endl;
		else if ( abs( x1 - x2 ) == abs( y1 - y2 )  )
			cout << 1 << ' ' << c1 << ' ' << y1 << ' '
			                 << c2 << ' ' << y2 << endl;
		else if ( y1 - y2 + x1 + x2 <= 14 && x1 - x2 + y1 + y2 <= 16 &&
		          y1 - y2 + x1 + x2 >= 0  && x1 - x2 + y1 + y2 > 0 )
			cout << 2 << ' ' << (char)( x1 + 'A' ) << ' ' << y1 << ' '
			                 << (char)( ( y1 - y2 + x1 + x2 ) / 2 + 'A' )
                             << ' '
			                 << ( x1 - x2 + y1 + y2 ) / 2 << ' '
			                 << (char)( x2 + 'A' ) << ' ' << y2 << endl;
		else
			cout << 2 << ' ' << (char)( x1 + 'A' ) << ' ' << y1 << ' '
			                 << (char)( ( y2 - y1 + x1 + x2 ) / 2 + 'A' )
                             << ' '
			                 << ( x2 - x1 + y1 + y2 ) / 2 << ' '
			                 << (char)( x2 + 'A' ) << ' ' << y2 << endl;
	}

	return 0;
}

File: Kattis/chess.py
alphabet = {'A': 0, 'B': 1, 'C': 2, 'D': 3, 'E': 4, 'F': 5, 'G': 6, 'H': 7}
reverse = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H']

cases = int(input())

for case in range(cases):

    x1, y1, x2, y2 = input().split()

    x1 = alphabet[x1]
    x2 = alphabet[x2]
    y1 = int(y1)
    y2 = int(y2)

    if (x1 + y1) % 2 != (x2 + y2) % 2:
        print("Impossible")
    elif x1 == x2 and y1 == y2:
        print("0", reverse[x1], y1)
    elif abs(x1 - x2) == abs(y1 - y2):
        print("1", reverse[x1], y1, reverse[x2], y2)
    elif 0 <= y1 - y2 + x1 + x2 and y1 - y2 + x1 + x2 <= 14 and \
            2 <= x1 - x2 + y1 + y2 and x1 - x2 + y1 + y2 <= 16:
        print("2", reverse[x1], y1,
              reverse[int((y1 - y2 + x1 + x2) / 2)],
              int((x1 - x2 + y1 + y2) / 2),
              reverse[x2], y2)
    else:
        print("2", reverse[x1], y1,
              reverse[int((y2 - y1 + x1 + x2) / 2)],
              int((x2 - x1 + y1 + y2) / 2),
              reverse[x2], y2)
